app/src/main/java/com/example/cinemapp/ui/main/HomeFragment.kt

```diff
@@ -0,0 +1,60 @@
+package com.example.cinemapp.ui.main
+
+import android.os.Bundle
+import android.util.Log
+import androidx.fragment.app.Fragment
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.recyclerview.widget.GridLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import com.example.cinemapp.databinding.FragmentHomeBinding
+import com.example.cinemapp.util.ifFalseToTrueThen
+import com.example.cinemapp.util.isEndOfScroll
+import com.example.cinemapp.util.observeFlowSafely
+import org.koin.androidx.viewmodel.ext.android.viewModel
+
+class HomeFragment : Fragment() {
+
+    private val viewModel by viewModel<HomeViewModel>()
+    private val adapter = MovieAdapter()
+
+    private var _binding: FragmentHomeBinding? = null
+    private val binding get() = _binding!!
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        _binding = FragmentHomeBinding.inflate(inflater, container, false)
+        viewModel.getUpcomingNextPage()
+        setupAdapter()
+        return binding.root
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        observeFlowSafely(viewModel.state) {
+            adapter.setMovies(it.movies)
+        }
+
+        binding.rvMovieList.addOnScrollListener(
+            object : RecyclerView.OnScrollListener() {
+                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
+                    super.onScrolled(recyclerView, dx, dy)
+                    ifFalseToTrueThen(
+                        viewModel.state.value.isTimeToPaginate,
```

Comment:
Fragment should not know how to access `viewModel.state.value.isTimeToPaginate` variable. Therefore, HomeViewModel should have api called something like `shouldPaginate()` to encapsulate this variable.